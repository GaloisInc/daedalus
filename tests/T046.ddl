
def l1 xs = for (s = 0 : int; x in xs) s + x
def l2 xs = for (s = 0 : int; k,x in xs) s + x + (k as int)

def ExampleMap = for (s = empty; k,x in [ 5 : int, 6, 7 ]) (Insert k x s)

def Main = {
  a1 = ^ l1 [1,2,3];
  a2 = ^ l2 [1,2,3];
  m  = ExampleMap;
  a3 = ^ l1 m;
  a4 = ^ l2 m;
}



