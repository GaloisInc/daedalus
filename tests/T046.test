T046.ddl
--interp=T046.ddl
