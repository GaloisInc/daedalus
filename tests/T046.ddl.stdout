module T046
 
--- Imports:
 
--- Type defs:
type T046.Main = { a1: int
                 ; a2: int
                 ; m: Map int int
                 ; a3: int
                 ; a4: int
                 }
 
--- Rules:
 
T046.ExampleMap : Grammar (Map int int) =
  for ((s : Map int int) = empty; (k : int), (x : int) in [5,
                                                           6,
                                                           7])
    (Insert k x s)
 
T046.l1 ?a0 ?a2 (Traversable ?a0) (HasKey ?a0 ?a2) (HasElement ?a0 int) (xs : ?a0) : int =
  for ((s : int) = 0; (x : int) in xs)
    (s + x)
 
T046.l2 ?a0 (Traversable ?a0) (HasKey ?a0 int) (HasElement ?a0 int) (xs : ?a0) : int =
  for ((s : int) = 0; (k : int), (x : int) in xs)
    ((s + x) + k)
 
T046.Main : Grammar T046.Main =
  do (a1 : int) <- pure (T046.l1 [int] int [1,
                                            2,
                                            3])
     (a2 : int) <- pure (T046.l2 [int] [1,
                                        2,
                                        3])
     (m : Map int int) <- T046.ExampleMap
     (a3 : int) <- pure (T046.l1 (Map int int) int m)
     (a4 : int) <- pure (T046.l2 (Map int int) m)
     pure {a1 = a1,
           a2 = a2,
           m = m,
           a3 = a3,
           a4 = a4}
 
T046._ExampleMap : Grammar (Map int int) =
  for ((s : Map int int) = empty; (k : int), (x : int) in [5,
                                                           6,
                                                           7])
    (Insert k x s)
 
T046._Main : Grammar {} =
  T046._ExampleMap
